### ISSUE TITLE
How the inplace_abn support multi-gpu and about the sequence of InPlaceABN operations?

### ISSUE THREAD

**Issue by PkuRainBow** [mapillary/inplace_abn#7]
Do we need to make some specific modifications to support multi-gpu training, in the conventional pytorch, we can simply set   **model = nn.DataParallel(deeplab)**.
So can we simply do this with inplace_abn?

Besides, I also meet the errors

_RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation_ 

I find that a guy @rotabulo said that 

> 
"you cannot have two subsequent inplace operations in your computation graph. E.g., if you use InPlaceABN, the preceding and subsequent layer should not be inplace."

But I find in your DenseNet implementations, you will have such operations like [InPlaceABN ->InPlaceABN -> InPlaceABN ...]. Which seems incorrect?

Could you help me?

**Comment by rotabulo** [comment]
@PkuRainBow you can use nn.DataParallel with models that have InplaceABN, indeed the imagenet training script does use it (see, https://github.com/mapillary/inplace_abn/blob/master/test_imagenet.py#L76).

You can of course have multiple InPlaceABN layers in the network, but not in direct composition. In DenseNet you never have InPlaceABN in direct composition. There is always a convolutional layer inbetween (https://github.com/mapillary/inplace_abn/blob/master/models/densenet.py#L48).

**Comment by John1231983** [comment]
Regarding to multiple gpus, I have a question about training batch norm. AFAIK, the tensorflow compute batch norm statistic in each GPU and used a BN statistics to update and for inference. That may not exploit the benefit of BN for training with larger of batch size. How about your method? Does it can compute BN statistic based on information of all GPUs? Thanks for great work

**Comment by rotabulo** [comment]
@John1231983 We have both implementations. InPlaceABN computes statistics locally in each GPU as in standard multi-gpu implementation of BN, while InPlaceABNSync (https://github.com/mapillary/inplace_abn/blob/master/modules/bn.py#L108) synchronizes the statistics across all the GPU during fwd and bwd passes. This makes the outcome equivalent to having computed the gradient for the entire batch on a single gpu. We report some experiments in the arxiv paper.

**Comment by John1231983** [comment]
Great feature. Do you have any plan to implement it in Tensorflow. I think many people included me will interest it
